host-macros/src/service.rs

```diff
@@ -137,32 +149,35 @@ impl ServiceBuilder {
             .enumerate()
             .map(|(index, args)| {
                 let name_screaming =
-                    format_ident!("DESC_{}_{index}", to_screaming_snake_case(characteristic.name.as_str()));
-                let properties = set_desc_access_properties(&args);
+                    format_ident!("DESC_{index}_{}", to_screaming_snake_case(characteristic.name.as_str()));
+                let access = &args.access;
+                let properties = set_access_properties(access);
                 let uuid = args.uuid;
-                // let read_callback = args.on_read.as_ref();
-                // let write_callback = args.on_write.as_ref();
-                let writeable = args.write || args.write_without_response;
+                let read_callback = match &access.on_read {
+                    Some(callback) => quote!(Some(#callback)),
+                    None => quote!(None),
+                };
+                let write_callback = match &access.on_write {
+                    Some(callback) => quote!(Some(#callback)),
+                    None => quote!(None),
+                };
+                let writeable = access.write || access.write_without_response;
+                let default_value = match &args.default_value {
+                    Some(val) => quote!(#val), // if set by user
+                    None => quote!(""),
+                };
                 let capacity = match &args.capacity {
                     Some(cap) => quote!(#cap),
                     None => {
                         if writeable {
                             panic!("'capacity' must be specified for writeable descriptors");
                         }
-                        quote!(0u8)
+                        quote!(#default_value.len() as u8)
                     },
                 };
-                let default_value = match &args.default_value {
-                    Some(val) => quote!(#val), // if set by user
-                    None => quote!(""),
-                };
-                // At least two attributes will be added to the attribute table for each characteristic:
-                // - The characteristic declaration
-                // - The characteristic's value declaration
-                //
-                // If the characteristic has either the notify or indicate property,
-                // a Client Characteristic Configuration Descriptor (CCCD) declaration will also be added.
-                self.attribute_count += if args.notify { 3 } else { 2 };
+
+                self.increment_attributes(access);
```

Comment:
@lulf

Reply:
>     1. if descriptors can be `notify` and/or `indicate`, do they also need a CCCD attribute to be pushed on to the table?

No, descriptors can't have descriptors (such as CCCD) and can't be notify or indicate.

>     2. currently this will also increment the attribute count by 2 / 3 for descriptors. Should we be only incrementing by 1 (or 2 with CCCD) for descriptors instead?

Yes, descriptors should only have a single attribute.

Reply:
Nice, I'll make that change, thanks!